src/silly_sprites/opengl/shader_program.cpp

```diff
@@ -0,0 +1,141 @@
+#include "shader_program.hpp"
+#include "utils.hpp"
+
+namespace sly::gl {
+    static constexpr char const* fallback_vertex_source{ R"(
+            #version 330 core
+            layout (location = 0) in vec3 aPos;
+
+            void main()
+            {
+                gl_Position = vec4(aPos.x, aPos.y, aPos.z, 1.0);
+            }
+        )" };
+    static constexpr char const* fallback_fragment_source{ R"(
+            #version 330 core
+            out vec4 FragColor;
+            
+            void main()
+            {
+                FragColor = vec4(255.0f, 0.0f, 255.0f, 1.0f);
+            }    
+        )" };
+
+    [[nodiscard]] constexpr std::string_view ShaderProgram::get_name_from_type(Type const type) {
+        switch (type) {
+            case Type::Vertex:
+                return "VERTEX";
+            case Type::Geometry:
+                return "GEOMETRY";
+            case Type::Fragment:
+                return "FRAGMENT";
+            default:
+                return "INVALID";
+        }
+    }
+
+    [[nodiscard]] ShaderProgram::Shader
+    ShaderProgram::compile(Type const type, std::string_view const source, bool const fallback) {
+        // compile
+
+        auto id{ glCreateShader(sly::to_underlying(type)) };
+        auto const* c_str = source.data();
+        glShaderSource(id, 1, &c_str, nullptr);
+        glCompileShader(id);
+
+        // check
+        GLint success;
+        glGetShaderiv(id, GL_COMPILE_STATUS, &success);
+        if (not success) {
+            GLchar message[512];
+            glGetShaderInfoLog(id, 512, nullptr, message);
+            spdlog::critical("ERROR::SHADER::{}::COMPILATION_FAILED -> {}", get_name_from_type(type), message);
+
+            if (not fallback) {
+                if (type != Type::Geometry) {
+                    glDeleteShader(id);
+                    spdlog::error("ERROR::SHADER::{}::FALLBACK", get_name_from_type(type));
+                    if (type == Type::Vertex) {
+                        return compile(type, fallback_vertex_source, true);
+                    } else { // Type::Fragment
+                        return compile(type, fallback_fragment_source, true);
+                    }
```

Comment:
Don't use a comment when you can use an assert.
```suggestion
                    } else {
                        assert(type == Type::Fragment);
                        return compile(type, fallback_fragment_source, true);
                    }
```